app/javascript/src/submit_handler.js

```diff
@@ -0,0 +1,108 @@
+
+const {
+  mergeObjects
+} = require('./utilities');
+
+class SubmitHandler {
+  #config;
+  #submitEnabled
+
+  constructor($form, config) {
+    this.#config = mergeObjects({
+      text: {
+        submitted: '',
+        unsubmitted: '',
+        submitting: '',
+      },
+      buttonSelector: ':submit',
+      buttonDescriptionSelector: '',
+      preventUnload: false,
+    }, config);
+    
+    this.$form = $form;
+    this.$button = this.$form.find(this.#config.buttonSelector);
+    this.text = this.#config.text;
+    if(this.#config.buttonDescriptionSelector) {
+      this.$buttonDescription = this.$form.find(this.#config.buttonDescriptionSelector);
+    }
+
+    this.$form.on('submit', (event) => this.#handleSubmit(event));
+    this.$button.on('click', (event) => this.#handleClick(event));
+  }
+
+  get required() {
+    return this.#submitEnabled;
+  }
+
+  set required(required) {
+    if(required) { 
+      this.#enable();
+    } else {
+      this.#disable();
+    }
+  }
+
+  forceSubmit() {
```

Comment:
Is there a reason this is called 'forceSubmit' rather than just 'submit'?
To me, the 'force' bit sounds like it is overriding something, or submitting regardless of any reason it shouldn't. 

Then again, since it looks like I could call `mySubmitHandler.forceSubmit()` directly in code, maybe that's the whole point. 

Should that be an option though? 

Makes me wonder if this introduces the possibility to (accidentally??) bypass the whole `this.required` checking in a potentially dangerous way. Was there a use-case or expectation driving a thought process here?

Reply:
Yep, if you check the PR description I explained this.  There is absolutely a rerquirement to be able to bypass the `required` functionality.

On a multiQ page, adding a new question or content area, triggers a relaod of the page.  If we don't bypass the `required` functionality then this fails.

So we either need to be able to do somehting like this from outside the class:
```
var tmp = submitHandler.enableSubmit;
submitHandler.enablesubmit = true
submitHandler.$form.submit();
submitHandler.enablesubmit = tmp
```

Which feels very messy, or we need a way for the class itself to be able to do this, which felt much neater... hence `forceSubmit()` or possibly `saveAndSubmit()`?  Although it doesn't save itself, just emits the event... I mean technically all it needs to do is remove the `window.unload` eventListeners and then trigger a submit.

Reply:
Ok, I missed that description text (note to self: always re-read the description!)
Also, I didn't notice the relevance of the change where you had...

```
$button.on("click.AddContent", () => {
      var submitEnabled = this.dataController.submitEnabled;
      updateHiddenInputOnForm(config.$form, fieldname, "content");
      this.dataController.submitEnabled = true;
      this.dataController.$form.submit();
      this.dataController.suibmitEnabled = submitEnabled;
    });
```

and updated it to...

```
$button.on("click.AddContent", () => {
      updateHiddenInputOnForm(config.$form, fieldname, "content");
      this.submitHandler.forceSubmit();
    });
```

I was going to mention in this comment that it need only have been changed to...

```
$button.on("click.AddContent", () => {
      updateHiddenInputOnForm(config.$form, fieldname, "content");
      this.submitHandler.required = true;
      this.submitHandler.submit(); 
    });
```

Since `this.submitHandler = true` already alters the value of private `enableSubmit` and, the resulting page reload would to the reset back to default `false` value. However, I can see now that the getter also activates the onunload handler, which would ruin the current action. Fair enough.

To address the this.submitHandler.forceSubmit() naming issue, I wondered if it could simply be...

```
$button.on("click.AddContent", () => {
      updateHiddenInputOnForm(config.$form, fieldname, "content");
      this.submitHandler.submit(true); 
    });
```

So that you could explicitly pass and optional value true (which presumably would change only the `enableSubmit` directly), thus indicating you knew why you were programatically submitting but then, especially given the use-case here, I'm wondering if either `submitHandler.forceSubmit()` or `submitHandler.submit(true)` (whatever variant) are actually necessary.

Maybe `submitHandler.submit()` is all that is required and the inner functionality actively handling the `enableSubmit` value is not a problem. The expected behaviour is that passing a Form (which finds the button) is enough to attach the required `SubmitHandler` functionality but, by simply providing a `submitHandler.submit()` method it can be assumed that it is safe to use - which it is, right? The `forceSubmit()` name, perhaps unnecessarily just implies there might be a reason to think twice or avoid using.

Reply:
I had also come to a similar conclusion that `submit()` was probably the best method name.